### DISCUSSION TITLE
Readability report generator

### DISCUSSION THREAD

**Discussion by mortii** [mortii/anki-morphs#37] [Ideas]
We need to decide on the desired input and output of this feature.

### Input
- In morphman it seems like you can only select directory, not just a single file in the directory? You should be able to do both imo.
- morphemizer drop down menu

### Output
This is the output morphman produces:
![cropped](https://github.com/mortii/anki-morphs/assets/15674619/be2d3f93-a972-4fdb-83df-70cc1537b036)

AnkiMorphs doesnt use the 'young' nor 'mature' terminology, only  unknown, learning, and known.
I don't know how I feel about the ' line readability %' and the 'i+1 lines %' columns, are they worth keeping?

Maybe there should be a toggle for either outputting percent or absolue values? E.g.
``` text
Total morphs - Unknown morphs - Learning morphs - Known morphs
```
and
``` text
Total morphs - Unknown morphs % - Learning morphs % - Known morphs %
```

If they are separated like that it would make the output less cluttered. Or maybe they can be in two different tabs, one output tab for absolute values, and another tab for percentages?

**Comment by Vilhelm-Ian** [comment]
I love the absolute number because it motivates me. In the sense that I have only 100 words left until I know all the words in the book. But honestly it's pointless. It dosen't help you in anyway in judging how much you know of the book.  The same thing about i+1 and line readability. They would be useful if a user is try to decide which book to read when both have approximatly the same readabilty but what is the chance of that happening

**Reply by mortii** [reply]
> I love the absolute number because it motivates me. In the sense that I have only 100 words left until I know all the words in the book. But honestly it's pointless. It dosen't help you in anyway in judging how much you know of the book. 

I would argue that if you love it then it is not pointless--motivation is very important.

How do you feel about one output tab for absolute numbers and another output tab for percentages?

**Reply by Vilhelm-Ian** [reply]
Much better than having it all in one place

**Comment by mortii** [comment]
Ok, I think the output should at  least have these columns:

Absolute numbers output tab:
```
Total Morphs
Unknown Morphs
Learning Morphs
Known Morphs
```

Percentage output tab:
```
Total Morphs
Unknown Morphs %
Learning Morphs %
Known Morphs %
```

Are the 'instances' columns interesting enough to include?

**Reply by Vilhelm-Ian** [reply]
No they are not

**Comment by mortii** [comment]
I made this UI so far:
![Screenshot from 2023-11-12 15-40-08](https://github.com/mortii/anki-morphs/assets/15674619/aee63317-5ec0-48c6-9b4c-f1ef41b6cd93)

I'm not sure what the 'select output' should be or do, but outputting to files and not just the tabs in the dialog is probably a useful thing no? Maybe it should be optional?

I also want to try and use a proper model-view-controller Qt setup for the tables in the tabs this time, unlike my more hacky approach in the settings dialog. Don't know how long that will take me to setup.

**Comment by mortii** [comment]
@Vilhelm-Ian what formats were the books you used in the readability analyzer? supporting epub would be sick, but I don't know how difficult that would be. IIRC epub is just html, so just stripping it might be fairly easy.

**Reply by aleksejrs** [reply]
I converted HTML files using `html2markdown.py3`, and PDF and epub files using Okular. md files I just renamed, because MorphMan looked at the extension. Subtitles are srt (common), ass (less common) and vtt (YouTube), some of which failed to process for some reason.

It should say if it encounters a file it will not process.

**Comment by Vilhelm-Ian** [comment]
.txt I copy pasted pdf files to txy

**Comment by mortii** [comment]
@aleksejrs @Vilhelm-Ian Like you guys pointed out there are a bunch of tools that can do the converting of the desired content, so we should just require the user to pre-process the files and make it txt or other equivalently simple formats, that way we don't have to worry about it.

**Comment by mortii** [comment]
I decided to reuse many of the parts from the frequency file generator. It will also be greedy (recursive) instead of single selection.

![Screenshot from 2023-12-11 14-57-46](https://github.com/mortii/anki-morphs/assets/15674619/07b41171-da94-4e93-b234-254d51aeb0df)

**Reply by Vilhelm-Ian** [reply]
I like the design.

**Reply by Vilhelm-Ian** [reply]
@mortii do you need help with the redability analyzer

**Reply by mortii** [reply]
@Vilhelm-Ian do you want to work on it? I can finish setting it up and then you could work on outputting stuff to the tabs maybe?

**Comment by Vilhelm-Ian** [comment]
> @Vilhelm-Ian do you want to work on it? I can finish setting it up and then you could work on outputting stuff to the tabs maybe?

I am busy with aoc currently. 

But yes I want to work on it. This is my favorite feature

**Reply by mortii** [reply]
@Vilhelm-Ian nice. I have pushed the code to the 'report' branch. I might have broken the frequency file generator in the refactoring process, but that can be fixed later.

**Reply by mortii** [reply]
@Vilhelm-Ian Have to made any progress on this? I'm waiting on a reply from the AnkiSpacy dev, and in the meantime I could probably finish a basic version of the readability report generator and release it.

**Reply by Vilhelm-Ian** [reply]
Noo. Advent of code is taking a lot more time time than I expected. Can you give me till monday

**Reply by mortii** [reply]
@Vilhelm-Ian sorry, I got too impatient, lol.

**Comment by mortii** [comment]
Finished, I'll update the docs and then push a new release:

![Screenshot from 2023-12-15 19-47-34](https://github.com/mortii/anki-morphs/assets/15674619/1d7ba8e7-f5a5-4627-9c9f-a774cf76ec39)

![Screenshot from 2023-12-15 19-47-30](https://github.com/mortii/anki-morphs/assets/15674619/71a45dbc-7e2d-461e-accb-1ca1cc2a2686)

**Reply by Vilhelm-Ian** [reply]
Thanks a lot. I have been waiting for this feature for ever. Sorry I couldn't help

**Reply by mortii** [reply]
@Vilhelm-Ian np np, i'm glad you can finally use this again :)

**Comment by Gardengul** [comment]
I would like to suggest memorizing the states of the flags. It is tedious to turn on the flags every time.
![image](https://github.com/mortii/anki-morphs/assets/95347489/68e7c8df-c093-415f-9850-ecb5ac74f562)

**Reply by mortii** [reply]
Agreed, it's on the todo-list ( #147) :+1:
